01_classify_stratified/nbconverted/plot_cancer_type_results.py

```diff
@@ -283,3 +285,136 @@ def label_points(x, y, gene, sig, ax):
+          size=14, pad=10)
+plt.xlabel('Cancer type', size=14)
+
+
+# ## Confusion matrix and clustering with subsampled cancer types
+# 
+# In the expression confusion matrix above, there are some strange "horizontal bars" (e.g. for PAAD and KICH) that we think are an artifact of label imbalance between cancer types of different sizes. To test this, we subsampled each cancer type down to the size of the smallest cancer type (~50 samples) and ran the same classification experiments, taking the average over 5 different random seeds to account for variability in the subsampling.
+
+# In[16]:
+
+
+preds_dir = os.path.join(cfg.repo_root, 'results_preds_subsampled', 'cancer_type')
+
+preds_expression_ss_df = au.load_preds_to_matrix(preds_dir, sample_info_df,
+                                                 training_data='expression')
+# take average over multiple seeds/subsampling replicates
+preds_expression_ss_df = preds_expression_ss_df.groupby(preds_expression_ss_df.index).mean()
+print(preds_expression_ss_df.shape)
+preds_expression_ss_df.iloc[:5, :5]
+
+
+# In[17]:
+
+
+sns.set({'figure.figsize': (15, 10)})
+ax = sns.heatmap(preds_expression_ss_df, cbar_kws={'label': 'Predicted probability of positive label, averaged over samples'})
+# this is needed to increase colorbar label font size
+# https://stackoverflow.com/a/48587137
+ax.figure.axes[-1].yaxis.label.set_size(14)
+plt.xlabel('True cancer type label', size=14)
+plt.ylabel('Positive label used to train classifier', size=14)
+plt.title('Cancer type confusion matrix, subsampled gene expression data', size=14, pad=14)
+plt.tight_layout()
+
+
+# In[18]:
+
+
+# fit agglomerative clustering model to confusion (similarity) matrix
+model = AgglomerativeClustering(distance_threshold=0,
+                                n_clusters=None,
+                                affinity='precomputed',
+                                linkage='complete')
+# takes a distance matrix (1 - similarity)
+model = model.fit(1 - preds_expression_ss_df.values)
+
+plot_dendrogram(model, labels=preds_expression_ss_df.index.values)
+plt.title('Dendrogram for complete-linkage hierarchical clustering, subsampled expression data',
+          size=14, pad=10)
+plt.xlabel('Cancer type', size=14)
+
+
+# The confusion matrix and clusters from the subsampled expression data look much more reasonable to us: the subsampling seems to get rid of the horizontal 'bars' from the non-subsampled analysis (fixes label imbalances), and most things that we expect to be clustered together are together (GBM/LGG, READ/COAD, SKCM/UVM, UCEC/UCS, DLBC/LAML). There are some idiosyncracies (KICH isn't with KIRC/KIRP, all are kidney cancers) but overall this seems like a reasonable starting point.
+# 
+# I also tried "smallest" and "average" linkage clustering - both give somewhat reasonable results, but "complete" linkage clustering gives the clearest separation between related cancer types, at least to my (relatively untrained) eye.
```

Comment:
What's the difference between these different linkings?

Reply:
Agglomerative hierarchical clustering essentially works by iteratively merging elements in a dataset (in this case, cancer types) based on some similarity or distance measure (in this case, confusion matrix probability) until a certain number of clusters are formed.

When you're comparing two individual elements/cancer types it's fairly obvious how this would work (just combine the most similar cancer types based on entries in the similarity matrix, then repeat with the next most similar, and so on), but eventually to form larget clusters you have to ask the question: how should we compare two clusters/groups of elements (or one cluster and a single element, etc)?

You can imagine lots of different ways to do this, but probably the most intuitive are:
* "Smallest" linkage clustering: distance between clusters = min distance between individual elements
* "Average" linkage clustering: distance between clusters = average distance between individual elements
* "Complete" linkage clustering: distance between clusters = max distance between individual elements

For whatever reason, complete linkage seems to work the best here. I don't have a great rationale for why that is, but my intuition is that it maximizes the effect of the worst-case comparisons -- that is, if there are a few very similar cancer types and one dissimilar cancer type, this will have a large complete linkage distance.

I don't think it would make a huge difference if I used average linkage, but smallest linkage does seem to give weird clusters in some cases.

Reply:
That logic makes sense to me